### ISSUE TITLE
Replace positional arguments with keyword arguments, merge CLI functions

### ISSUE THREAD

**Issue by valgur** [sentinelsat/sentinelsat#115]
I propose we replace the CLI positional arguments `<user> <password> <geojson>` with equivalent "optional" arguments (such as `-u,--user`, `-p,--password` and `-g,--geometry`). The `user` and `password` args would still be required arguments in one form or another, of course.

Here are some of the possible benefits of that change:
* It would allow us to implement alternative ways of passing credentials to our CLI, e.g. through [environment variables](http://click.pocoo.org/5/options/#dynamic-defaults-for-prompts), INI or .netrc files (#90).
* Would enable alternative arguments for passing AOI geometries, if need be (#114).
* The user does not need to remember the order of the positional args and can pass arguments in any order. A minor convenience.
* One of ours goals for release v1.0 should be stability and backwards-compatibility of the CLI and API (#108). So now would be a good time for this backwards-incompatible change. This would provide greater flexibility for us in the future, though.

Do you agree?

**Comment by Fernerkundung** [comment]
Completely agree on the proposed changes. I would also add:

- merge `sentinel search` and `sentinel download` into `sentinelsat`

The download of a single scene is the only function of `download`, which can be included into a unified CLI quite easily as a positionaly argument such as `-i, --id`.

**Comment by j08lue** [comment]
I strongly support both of these suggestions.

**Comment by valgur** [comment]
Good idea, @Fernerkundung. I suggest using just `--id` as the keyword, since that's short and consistent with the server-side name. I would also add `--title` to search by the product titles ("S1A_IW_GRDH_...") as well.

**Comment by Fernerkundung** [comment]
I would suggest using `--id` as shorthand for the `UUID` and `--name` to refer `product name` / `file name`, to be consistent with the [OpenHub doumentation](https://scihub.copernicus.eu/twiki/do/view/SciHubUserGuide/5APIsAndBatchScripting#Filtering_the_Products_using_the)

**Comment by valgur** [comment]
I have an alternative idea for the CLI setup: we can keep the `search` and `download` commands but make them share the search parameter arguments. I googled a bit and this can be done pretty easily: https://github.com/pallets/click/issues/108
It would keep the commands and also each commands arguments separate, so arguments that are only relevant to downloading don't pollute the list of arguments for searching, for example. Seems a lot tidier to me.

**Comment by j08lue** [comment]
So you will use something along the lines of [this answer](https://github.com/pallets/click/issues/108#issuecomment-255547347)?

Because `click` did not get any high-level support for this, did it?

**Comment by valgur** [comment]
I liked [this one](https://github.com/pallets/click/issues/108#issuecomment-280489786) a bit more, but pretty much, yes.

**Comment by Fernerkundung** [comment]
That is a nice solution to keeping `search` and `download` separate. However, I am still strongly in favour of merging both into a unified CLI command.

- the CLI should be the easy to use entry point so we have to aim to keep it simple
- `download` only fulfills the function of downloading a single scene, which could be merged into `search` or a unified CLI
- `sentinel search` is descriptive but it could cause confusion why the first part of the CLI command is `sentinel` instead of the name of the tool `sentinelsat`
- download of a single scene is often fastest dealt with by using the DHuS GUI - therefore I don't expect that `download` sees a lot of use

**Comment by valgur** [comment]
Oh. I should have been clearer. What I had in mind was changing `sentinel download` to work as `sentinel search --download` works right now.
So the `search` command would look like
```
Usage: sentinel search [OPTIONS] <user> <password> <geojson>

  Search for Sentinel products and, optionally, create a geojson file with the 
  search result footprints.

Options:
  -s, --start TEXT                Start date of the query in the format
                                  YYYYMMDD.
  -e, --end TEXT                  End date of the query in the format
                                  YYYYMMDD.
  -f, --footprints                Create a geojson file
                                  search_footprints.geojson with footprints
                                  and metadata of the returned products.
  -q, --query TEXT                Extra search keywords you want to use in the
                                  query. Separate
                                  keywords with comma.
                                  Example:
                                  'producttype=GRD,polarisationmode=HH'.
  -u, --url TEXT                  Define another API URL. Default URL is
                                  'https://scihub.copernicus.eu/apihub/'.
  --sentinel [1|2|3]              Limit search to a Sentinel satellite
                                  (constellation)
  --instrument [MSI|SAR-C SAR|SLSTR|OLCI|SRAL]
                                  Limit search to a specific instrument on a
                                  Sentinel satellite.
  --producttype [SLC|GRD|OCN|RAW|S2MSI1C|S2MSI2Ap]
                                  Limit search to a Sentinel product type.
  -c, --cloud INTEGER             Maximum cloud cover in percent. (requires
                                  --sentinel to be 2 or 3)
  -o, --order-by TEXT             Comma-separated list of keywords to order
                                  the result by. Prefix keywords with '-' for
                                  descending order.
  -l, --limit INTEGER             Maximum number of results to return.
                                  Defaults to no limit.
  --help                          Show this message and exit.
```
and for `download`:
```
Usage: sentinel download [OPTIONS] <user> <password> <geojson>

  Download Sentinel products returned from a search.

Options:
  -s, --start TEXT                Start date of the query in the format
                                  YYYYMMDD.
  -e, --end TEXT                  End date of the query in the format
                                  YYYYMMDD.
  -p, --path PATH                 Set the path where the files will be saved.
  -q, --query TEXT                Extra search keywords you want to use in the
                                  query. Separate
                                  keywords with comma.
                                  Example:
                                  'producttype=GRD,polarisationmode=HH'.
  -u, --url TEXT                  Define another API URL. Default URL is
                                  Limit search to Sentinel-2 products.
  --sentinel [1|2|3]              Limit search to a Sentinel satellite
                                  (constellation)
  --instrument [MSI|SAR-C SAR|SLSTR|OLCI|SRAL]
                                  Limit search to a specific instrument on a
                                  Sentinel satellite.
  --producttype [SLC|GRD|OCN|RAW|S2MSI1C|S2MSI2Ap]
                                  Limit search to a Sentinel product type.
  -c, --cloud INTEGER             Maximum cloud cover in percent. (requires
                                  --sentinel to be 2 or 3)
  -o, --order-by TEXT             Comma-separated list of keywords to order
                                  the result by. Prefix keywords with '-' for
                                  descending order.
  -l, --limit INTEGER             Maximum number of results to return.
                                  Defaults to no limit.
  --help                          Show this message and exit.
```

**Comment by Fernerkundung** [comment]
I understood what you had in mind, but I still favour the premise of this PR to "Replace positional arguments with keyword arguments" - your proposition would keep download as a positional argument, whereas I'd rather it be a keyword argument as well and search / query is the default behaviour if only to produce less traffic in case of user error.

**Comment by j08lue** [comment]
I would not call `download` in the current setup a positional argument, but a subcommand.

I agree with you, @Fernerkundung, that our CLI is much clearer with just one `sentinelsat` command to search scenes and with a `--download` option to quickly turn that search into files.

We also do not gain much simplicity by keeping the `download`-related commands out of the `search` signature, because there are so few that are not common to the two functions.

**Comment by valgur** [comment]
I agree with all of the points you make. `--download` is the way to go. Thanks for the discussion. 👍

**Comment by valgur** [comment]
Resolved by #143.
